microscopium/cluster.py

```diff
@@ -161,3 +163,123 @@ def mds_mapping(X, n_components=2, max_iter=500, n_jobs=-1,
+        pass
+
+    def _add_first_sample(self, x):
+        self._online_mean = np.array(x, dtype=np.float)
+        self._online_sq_mean = np.zeros_like(self._online_mean)
+        self._n += 1
+        self.add_sample = self._add_sample
+
+    def _add_sample(self, x):
+        x = np.asanyarray(x)
+        self._n += 1
+        mean, sq_mean, n = self._online_mean, self._online_sq_mean, self._n
+        delta = x - mean
+        mean += delta / n
+        sq_mean += delta * (x - mean)
+
+    def mean(self):
+        """Return the current mean."""
+        return np.array(self._online_mean, copy=True)
+
+    def var(self):
+        """Return the current variance."""
+        return self._online_sq_mean / self._n
+
+    def standard_scaler(self):
+        """Return a sklearn.preprocessing.StandardScaler"""
+        s = StandardScaler()
+        s.mean_ = self.mean()
+        var = self.var()
+        var[var <= 0] = 1  # ignore variables with zero variance
+        s.std_ = np.sqrt(var)
+        return s
+
+
+class OnlineIncrementalPCA(object):
+    """Object to stream over a dataset and perform incremental PCA.
+
+    Parameters
+    ----------
+    n_components : int, optional
+        The number of PCA components to keep.
+    whiten : bool, optional
+        Whether to whiten the input. A good idea when input scales
+        vary widely.
+    batch_size : int, optional
+        The batch_size to use for the computation. (Streamed elements
+        are accumulated until this batch size is reached.)
+    """
+    def __init__(self, n_components=None, whiten=True, batch_size=None):
+        self.ipca = IncrementalPCA(n_components=n_components, whiten=whiten,
+                                   copy=True, batch_size=batch_size)
+        self.current_batch = []
+        self.batch_size = batch_size
+
+    def add_sample(self, v):
+        """Add a new sample to the model being learned.
+
+        These samples are "stocked up" until the current stock matches
+        the IPCA batch size (set at creation time). If the total number
+        of samples doesn't divide cleanly into the batch size, the
+        remainder of the samples will not be fit. (But that should be
+        fine, really!)
+
+        Parameters
+        ----------
+        v : array of float
+            The sample to add to the fit.
+        """
+        self.current_batch.append(np.squeeze(v))
+        if len(self.current_batch) >= self.batch_size:
+            self.ipca.partial_fit(np.array(self.current_batch))
+            del self.current_batch
+            self.current_batch = []
+
+    def transform(self, v):
+        """Transform `v` from the full space to the decomposed space...
+
+        ... according to the current IPCA model.
+
+        Parameters
+        ----------
+        v : array of float
+            The vector to transform.
+
+        Returns
+        -------
+        comp : array of float
+            The vector in the IPCA's reduced/rotated space.
```

Comment:
You are correct! I'll change this docstring. It reflects my intended use-case (streaming feature vectors), rather than the actual capabilities of the function.